http_server.go

```diff
@@ -0,0 +1,51 @@
+package main
+
+import (
+	"context"
+	"log"
+	"net"
+	"net/http"
+	neturl "net/url"
+	"time"
+)
+
+func startHTTPServer(ctx context.Context, handler http.Handler, logger *log.Logger) (url string, shutdown context.CancelFunc, err error) {
+	// Need to generate a random port every time for tests in parallel to run.
+	l, err := net.Listen("tcp", "localhost:")
+	if err != nil {
+		return "", nil, err
+	}
+
+	server := &http.Server{
+		Handler: handler,
+	}
+	go func() { // serves HTTP
+		err := server.Serve(l)
+		if err != http.ErrServerClosed {
+			logger.Println(err)
+		}
+	}()
+
+	shutdownCtx, shutdownNoWait := context.WithCancel(ctx)
```

Comment:
Why is this `shutdownNoWait`? The `shutdownComplete` channel waits for the shutdown to complete.

Reply:
The meaning probably isn't very clear. I was going for "shutdown without the added timeout attached", but perhaps `startShutdown` is clearer? Updated in 04568e5